server/etcdserver/api/membership/cluster.go

```diff
@@ -542,6 +543,53 @@ func (c *RaftCluster) PromoteMember(id types.ID, shouldApplyV3 ShouldApplyV3) {
 	)
 }
 
+func (c *RaftCluster) SyncLearnerPromotionIfNeeded() {
+	c.Lock()
+	defer c.Unlock()
+
+	v2Members, _ := membersFromStore(c.lg, c.v2store)
+	v3Members, _ := membersFromBackend(c.lg, c.be)
+
+	for id, v2Member := range v2Members {
+		v3Member, ok := v3Members[id]
+
+		if !ok {
+			c.lg.Error("Detected member only in v2store but missing in v3store", zap.String("member", fmt.Sprintf("%+v", *v2Member)))
+			continue
+		}
+
+		// A peerURL list is considered the same regardless of order.
+		clonedV2Member := v2Member.Clone()
+		sort.Strings(clonedV2Member.PeerURLs)
+
+		clonedV3Member := v3Member.Clone()
+		sort.Strings(clonedV3Member.PeerURLs)
+
+		if reflect.DeepEqual(clonedV2Member.RaftAttributes, clonedV3Member.RaftAttributes) {
+			c.lg.Info("Member's RaftAttributes is consistent between v2store and v3store", zap.String("member", fmt.Sprintf("%+v", *v2Member)))
+			continue
+		}
+
+		// Sync member iff both conditions below are true,
+		//  1. IsLearner is the only field in RaftAttributes that differs between v2store and v3store.
+		//  2. v2store.IsLearner == false && v3store.IsLearner == true.
+		clonedV3Member.IsLearner = false
+		if reflect.DeepEqual(clonedV2Member.RaftAttributes, clonedV3Member.RaftAttributes) {
+			syncedV3Member := v3Member.Clone()
+			syncedV3Member.IsLearner = false
+			c.lg.Warn("Syncing member in v3store", zap.String("member", fmt.Sprintf("%+v", *syncedV3Member)))
+			unsafeSaveMemberToBackend(c.lg, c.be, syncedV3Member)
```

Comment:
`unsafeSaveMemberToBackend` uses `LockInsideApply` this means two thing:
* This would cause CI to be incorrectly flushed (see my recommendation above about doing changes to members from apply loop).
* The validation for `LockInsideApply` didn't catch this locking. We should double check why.

Reply:
> * This would cause CI to be incorrectly flushed (see my recommendation above about doing changes to members from apply loop).

thx for pointing this out. Will update to use the LockOutsideApply.



> * The validation for `LockInsideApply` didn't catch this locking. We should double check why.

because there isn't an e2e test to trigger it being executed. 

PTAL https://github.com/etcd-io/etcd/pull/19602, once it gets merged, will continue to update that e2e test in this PR.

Reply:
updated.